### DISCUSSION TITLE
Generic of Generic Types contraints (chained generic constraints)

### DISCUSSION THREAD

**Discussion by redknightlois** [dotnet/csharplang#1612] [Language Ideas]
Last few years with all the investment made at optimizing value types at the JIT level it has opened a new design window that didn't exist before. I have seen and used myself lots of tricks to allow metaprogramming and zero cost extensions using interfaces and structs (very related to #110). 

While the case of Generic of Generic constraint is orthogonal to #110, it is very related to it. There are a few corner cases that I haven't seen mentioned in any of the concepts proposals but has a high impact on the code you can write (but more importantly to use). If I am wrong and it is already considered feel free to close the issue.

The case goes like this:
```csharp
public interface IAllocatorOptions {}

public unsafe interface IAllocator<T, TConfig> 
     where T : struct, IAllocator, IDisposable
     where TConfig : struct, IAllocatorOptions;
{
    int Allocated { get; }
    void Initialize(ref T allocator);
    void Configure(ref T allocator, ref TConfig configuration)
    void Allocate(ref T allocator, int size, out BlockPointer.Header* header);
    void Release(ref T allocator, in BlockPointer.Header* header);
    void Reset(ref T allocator);
}

public sealed class Allocator<TAllocator, TAllocatorConfiguration> : IDisposable
       where TAllocator: struct, IAllocator<TAllocator, TAllocatorConfiguration>
       where TAllocatorConfiguration: struct, IAllocatorOptions;
{}     
```

Problem is now your usage goes like this:
```csharp
var alloc = new Allocator<NativeBlockAllocator<NativeBlockAllocator.Secure>, NativeBlockAllocator.Secure>()
```

Chaining is becoming more prevalent as we are writing concepts by hand, but would also impact generic concept types. This could be easily be solved without any runtime change with something like this:
```csharp
public sealed class Allocator<TAllocator> : IDisposable
       where TAllocator: struct, IAllocator<TAllocator, generic TAllocatorConfiguration>
       where TAllocatorConfiguration : struct, IAllocatorOptions
{}     
```

Now the actual usage would go like this:
```csharp
var alloc = new Allocator<NativeBlockAllocator<NativeBlockAllocator.Secure>>()
```

With the upside that because it is kind-of a better type inference mechanism for generic types, it opens a whole new avenue for writing very customizable and efficient generic code through metaprogramming.

**Comment by ufcpp** [comment]
more complex case: https://github.com/ufcpp/KdTree/blob/7864b50ec5c1ea61d27674d5bc90282215b21f81/KdTreeTestsLib/KdTreeTests.cs#L5

```cs
using Tree = KdTree.KdTree<float, string, KdTree.Fixed2<float>.Array, KdTree.Fixed2<float>, KdTree.Math.FloatMath, KdTree.Math.EuclideanMetic<float, KdTree.Fixed2<float>.Array, KdTree.Fixed2<float>, KdTree.Math.FloatMath>>;
```

This is 2-dimensional fixed array of float with Euclidean metrics.

**Comment by redknightlois** [comment]
Holy moly that is a huge one :D

**Comment by Joe4evr** [comment]
Relevant reading: https://ericlippert.com/2013/07/15/why-are-generic-constraints-not-inherited/

**Comment by svick** [comment]
I *think* this might be effectively a combination of "Improved inference from generic constraints" https://github.com/dotnet/csharplang/issues/741 and "Partial type inference" https://github.com/dotnet/csharplang/issues/1349.

**Comment by redknightlois** [comment]
@Joe4evr Not really, we are talking about 2 different beasts here. 

This is generic chaining, which in itself is far more restricted that what Eric is talking about (as far as I can understand). Which is a type of higher degree generic type. Generic chaining can be reduced (unless I am missing something) to something that we already describe on the language itself through a very verbose notation. There is no type substitution with proper higher-order inference required. 

Say we use my example:
```csharp
public sealed class Allocator<TAllocator> : IDisposable
       where TAllocator: struct, IAllocator<TAllocator, generic TAllocatorConfiguration>
       where TAllocatorConfiguration : struct, IAllocatorOptions
{}     
```

This is a shortcut for (which is the actual code I use):
```csharp
public sealed class Allocator<TAllocator, TAllocatorConfiguration> : IDisposable
       where TAllocator: struct, IAllocator<TAllocator, TAllocatorConfiguration>
       where TAllocatorConfiguration : struct, IAllocatorOptions
{}     
```

There is nothing in it that would require substitution. 

> When you construct a generic type, you are required to prove to the compiler that the type you’ve supplied meets the constraints. In the case of generic type parameters that are themselves supplied as type arguments, you provide that proof by constraining them explicitly, not relying on the compiler to make a guess at what you meant.

By the time the definition happens everything is already defined. Now the question is what would happen at the call site? Which looks like this:
```csharp
var alloc = new Allocator<NativeBlockAllocator<NativeBlockAllocator.Secure>>()
```

Now here because this is a chained generic type (not a higher-order type), the actual generated code would look like this:
```csharp
var alloc = new Allocator<NativeBlockAllocator<NativeBlockAllocator.Secure>, NativeBlockAllocator.Secure>()
```
Again all the information necessary for the compiler at compilation time is already there, the constraints are in place already. 

Sure it syntactic sugar, but what such a facility addresses is the consumer code verbosity.

**Comment by redknightlois** [comment]
@svick Probably. I just don't see how those 2 would play together to solve this. The examples are mostly on shallow types, couldn't find any example that would give me an idea on how chained ones would work.

**Comment by xoofx** [comment]
Indeed, that's a long standing issue, bitting us quite often when we try do abuse generic. It is not full HKT as I said earlier, so it sounds reasonable. 

Though, for the compiler folks, not sure it will be an acceptable burden for the compiler type/override resolution...

The problem is that when you declare:

```c#
public sealed class Allocator<TAllocator> : IDisposable
       where TAllocator: struct, IAllocator<TAllocator, generic TAllocatorConfiguration>
       where TAllocatorConfiguration : struct, IAllocatorOptions
{}     
```

It is implicitly declaring something like `Allocator<TAllocator, TAllocatorConfiguration>`

So you should still be allowed to declare:

```c#
public sealed class Allocator<TAllocator> : IDisposable
       where TAllocator: struct, IAnotherAllocator
{}     
```

1. So the compiler will first have to make sure that the two types are different (which in the past, it just had to check the number of generic arguments)... 

2. Also when the compiler will see something like `new Allocator<MyAllocator>()` it will have to match against multiple potential candidates (while in the past it had at to match to only one) and resolve them more deeply (ala overrides)

For solving 1), the following syntax by using the keyword `implicit` might give less trouble, as we are explicitly declaring up-front the number of generic parameters and that the remaining parameter is implicit:

```c#
public sealed class Allocator<TAllocator, implicit TAllocatorConfiguration> : IDisposable 
    where TAllocator: struct, IAllocator<TAllocator, TAllocatorConfiguration>
    where TAllocatorConfiguration : struct, IAllocatorOptions
{
}
```

The good thing is that it can translate to IL to use an attribute instead:

```c#
    public class ImplicitAttribute : Attribute
    {
    }
    public sealed class Allocator<TAllocator, [Implicit] TAllocatorConfiguration> : IDisposable 
```

Not sure I would allow a non implicit generic after an implicit one to avoid holes in the declaration (e.g `class Allocator<T1, implicit T2, T3, implicit T4>`)

After for 2), I don't know how much trouble it is to have multiple generic type candidates but considering how much overrides resolution is sensitive (resolution wise, perf wise...), that's likely not going to be easy for this feature to go through...
